lxd/storage/drivers/driver_dir_utils.go

```diff
@@ -12,6 +12,15 @@ import (
 	"github.com/lxc/lxd/shared/units"
 )
 
+// withoutGetVolID returns a copy of this struct but without the volId property set.
+func (d *dir) withoutGetVolID() Driver {
+	newDriver := &dir{}
+	newDriver.init(d.state, d.name, d.config, d.logger, nil, d.getCommonVolumeRules)
```

Comment:
@stgraber I'm concerned that the approach used here, of passing a nil getIDFunc to newDriver.init(), and then the associated changes in the driver dir to make it tolerate (and ignore) the getIDFunc when it is nil, makes it too easy in the future to forget to provide this function and then volumes will have their quotas silently not setup (rather than crashing hard with a clear stack trace).

I understand why you're doing this though, as I believe you want to re-use the volume creation functions from the driver dir for backup restoration, which occurs before the DB records have been created, and thus no volume ID is available for quota setup.

So how about this?

Firstly, define the special VolID value `-1` as a constant, such as `VolIDQuotaSkip`. That way the use and meaning of its value across the dir driver can be easily linked and understood.

Secondly, rather than passing a nil function to this copy of the driver (and making the driver tolerate the nil value internally), instead pass a temporary function that just returns the `VolIDQuotaSkip` value always as the VolID.

That way it is explicit and non-accidental that we are passing a special function that skips quota, and we don't have to make the dir driver tolerant of a nil volIDFunc.

We'd then end up with just checks for  `if volID == VolIDQuotaSkip {` rather than `if volID == -1 {` which is understand.

Reply:
fixed